ocis-pkg/sync/mutex.go

```diff
@@ -0,0 +1,74 @@
+package sync
+
+import (
+	"sync"
+)
+
+// NRWMutex works the same as RWMutex, the only difference is that it stores mutexes in a map and reuses them.
+// It's handy if you want to write-lock, write-unlock, read-lock and read-unlock for specific names only.
+type NRWMutex struct {
```

Comment:
What would the impact be of, as you mention in your comment, having a type like:

```golang
type MapSemaphores map[string]struct{
	m sync.RWMutex
	c int
}
```

This way we could also spare the `nrw` private type.

Reply:
I'm aware of the top-level mutex, but still trying to wrap my head around why is it needed. The way I see it, having a map with a list of `[string]RWMutex`, having a semaphore as part of the `struct` leads me to believe that the `[string]RWMutex` member of the `struct` is part of the critical section, so many goroutines don't try and change stuff inside there. But the thing is that the data structure is just a bunch of semaphores, so why closing access to the entire critical section, when inside the critical section there are actually more mutexes, preventing goroutines from changing yet a deeper critical section, which as it turns out, is probably the real data (in this case, an account file).

Reply:
sorry if it's too confusing, I'm slowly digesting the code :D

Reply:
Thanks for your input.

Yes, that confused me too, we are not able to guarantee that 2 go routines won't ask for the same named lock in order. In this scenario it's possible that routine b tries to lock an already locket mutex. the same can happen to unlocking.

In my tests (https://github.com/owncloud/ocis/blob/account-locking/ocis-pkg/sync/mutex_test.go) i have found out that it's also important that the main lock needs to be unlocked after the current lock unlocks and for a lock of current lock the main lock needs to be unlocked before the current lock unlocks.

It's a bit complex :(

Reply:
> we are not able to guarantee that 2 go routines won't ask for the same named lock in order


@fschade so we have the issue that 2 goroutines are codependent? Say `b` wants to read data that `a` will write, but if `b` runs before `a` then we have a problem; could that be the case? 🤔

Reply:
let me try to explain it by example

```go
m := NewNRWMutex()
c := make(chan bool)

for i := 0; i < 500; i++ {
	go func(ii int, cc chan bool) {
		m.Lock(strconv.Itoa(ii))
		time.Sleep(500 * time.Millisecond )
		fmt.Printf("a - %v\n", ii)
		m.Unlock(strconv.Itoa(ii))
		c <- true
	} (i, c)
}

for i := 0; i < 500; i++ {
	go func(ii int, cc chan bool) {
		m.Lock(strconv.Itoa(ii))
		time.Sleep(time.Duration(rand.Int63n(100)) * time.Millisecond )
		fmt.Printf("b - %v\n", ii)
		m.Unlock(strconv.Itoa(ii))
		c <- true
	} (i, c)
}

for i := 0; i < 500; i++ {
	go func(ii int, cc chan bool) {
		m.Lock(strconv.Itoa(ii))
		fmt.Printf("c - %v\n", ii)
		m.Unlock(strconv.Itoa(ii))
		c <- true
	} (i, c)
}

for i := 0; i < 1500; i++ {
	<-c
}
```


without locks on NRWMutex:
- we have concurrent writes to the underlying map of mutexes (concurrent map writes errors)
- even if we use a sync map here, we cannot guarantee that a unlocks the same pointer mutex before b tries to lock it again

the result for example can be mixed and will conflickt without main locking (10 a, 10 b, 10 c):
```bash
c - 9
c - 7
c - 8
b - 8
a - 0
a - 9
a - 7
a - 2
a - 5
a - 4
c - 5
a - 1
a - 6
a - 3
c - 2
c - 4
c - 6
c - 3
a - 8
b - 6
b - 9
b - 1
c - 1
b - 5
b - 2
b - 3
b - 7
b - 0
c - 0
b - 4
```

as an alternative we could use the atomic package, but this is unnecessary complex

Reply:
> > we are not able to guarantee that 2 go routines won't ask for the same named lock in order
> 
> @fschade so we have the issue that 2 goroutines are codependent? Say `b` wants to read data that `a` will write, but if `b` runs before `a` then we have a problem; could that be the case? 🤔
> 
> -- sorry i edited it, i wanted to reply

in the performance tests yes and under high load maybe yes too (3 clients same user, one on cell, one on computer and another one with the web client). Then concurrent read writes can happen. For example in the indexer. Here i uses the typeName as a key for the named mutex

Reply:
> even if we use a sync map here, we cannot guarantee that a unlocks the same pointer mutex before b tries to lock it again

How come we cannot guarantee that a unlocks the same pointer mutex? :o [I toyed around with a little something](https://gist.github.com/refs/fd5d23aef913fe456851689340e210da), I think that can safely store and retrieve on a `sync.Map` pointers to `sync.RWMutex` 🤔. The rest of the implementation on my gist are nuances I was messing with.

Reply:
you can consistently retrieve mutexes by loading them from the `sync.Map`, for instance:

```golang
func TestLocks(t *testing.T) {
	d := &db{
		mutexes: sync.Map{},
		values: sync.Map{},
	}

	for i := 0; i < _nRoutines; i++ {
		d.writeMutex(strconv.Itoa(i), &sync.RWMutex{})
	}

	// lock mutex[1]
	// attempt to lock mutex[1] again
	mOne, _ := d.mutexes.Load(strconv.Itoa(1))
	mOne.(*sync.RWMutex).Lock()

	mOneB, _ := d.mutexes.Load(strconv.Itoa(1))
	mOneB.(*sync.RWMutex).Lock() // 💥
}
```

fails because I try to lock a mutex that is already locked.

Reply:
yes that is true, but if you compare the lock and unlock of the rwmutex you can see that:
* for lock the main lock gets released BEFORE the current lock gets locked
* for unlock the main lock gets released AFTER the current lock gets unlocked

in case of sync.map it gets released in booth cases BEFORE we lock or unlock the current lock. But i will be happy if find another way to get rid of it and keep the test green ;)

Reply:
here is the version with a sync map, the access to the map is atomic, but not the locking

https://gist.github.com/fschade/6117c564333138c4fbc0ef129f873a8b

```bash
=== RUN   TestNewNRWMutex
fatal error: sync: Unlock of unlocked RWMutex

goroutine 23 [running]:
runtime.throw(0x124f32b, 0x20)
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/panic.go:1116 +0x72 fp=0xc00004de88 sp=0xc00004de58 pc=0x1038472
sync.throw(0x124f32b, 0x20)
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/panic.go:1102 +0x35 fp=0xc00004dea8 sp=0xc00004de88 pc=0x106a5b5
sync.(*RWMutex).Unlock(0xc0000161a0)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/rwmutex.go:129 +0xc6 fp=0xc00004dee0 sp=0xc00004dea8 pc=0x1079e86
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Unlock(0xc0001cf560, 0x13a0b00, 0x1)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:34 +0xba fp=0xc00004df58 sp=0xc00004dee0 pc=0x11e437a
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:13 +0xd7 fp=0xc00004dfc8 sp=0xc00004df58 pc=0x11e5297
runtime.goexit()
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc00004dfd0 sp=0xc00004dfc8 pc=0x106f181
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000106c00, 0x12497c9, 0xf, 0x125a9b8, 0x108ba01)
	/usr/local/Cellar/go/1.15.3/libexec/src/testing/testing.go:1169 +0x2da
testing.runTests.func1(0xc000106480)
	/usr/local/Cellar/go/1.15.3/libexec/src/testing/testing.go:1439 +0x78
testing.tRunner(0xc000106480, 0xc000173de0)
	/usr/local/Cellar/go/1.15.3/libexec/src/testing/testing.go:1123 +0xef
testing.runTests(0xc00012a4e0, 0x13c2040, 0x4, 0x4, 0x0, 0x0, 0x0, 0x100f390)
	/usr/local/Cellar/go/1.15.3/libexec/src/testing/testing.go:1437 +0x2fe
testing.(*M).Run(0xc00017e080, 0x0)
	/usr/local/Cellar/go/1.15.3/libexec/src/testing/testing.go:1345 +0x1eb
main.main()
	_testmain.go:49 +0x138

goroutine 22 [chan receive]:
github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex(0xc000106c00)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:31 +0x12b
testing.tRunner(0xc000106c00, 0x125a9b8)
	/usr/local/Cellar/go/1.15.3/libexec/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/usr/local/Cellar/go/1.15.3/libexec/src/testing/testing.go:1168 +0x2b3

goroutine 24 [semacquire]:
sync.runtime_SemacquireMutex(0xc0001cf56c, 0xc000037600, 0x1)
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc0001cf568)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:81
sync.(*Map).LoadOrStore(0xc0001cf568, 0x1201100, 0xc0001cd640, 0x1220320, 0xc000146920, 0x1, 0x1, 0x13a0b00)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/map.go:209 +0x60e
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).loadOrStore(...)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:59
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Lock(0xc0001cf560, 0x13a0b00, 0x1)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:26 +0x89
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:12 +0xb5
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 25 [runnable]:
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:11 +0x3e
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 26 [semacquire]:
sync.runtime_SemacquireMutex(0xc0001cf56c, 0xc000038600, 0x1)
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc0001cf568)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:81
sync.(*Map).LoadOrStore(0xc0001cf568, 0x1201100, 0xc0001cd650, 0x1220320, 0xc000146940, 0x1, 0x1, 0x13a0b00)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/map.go:209 +0x60e
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).loadOrStore(...)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:59
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Lock(0xc0001cf560, 0x13a0b00, 0x1)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:26 +0x89
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:12 +0xb5
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 27 [semacquire]:
sync.runtime_SemacquireMutex(0xc0001cf56c, 0xc000038e00, 0x1)
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc0001cf568)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:81
sync.(*Map).LoadOrStore(0xc0001cf568, 0x1201100, 0xc0001cd660, 0x1220320, 0xc000146960, 0x1, 0x1, 0x13a0b00)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/map.go:209 +0x60e
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).loadOrStore(...)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:59
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Lock(0xc0001cf560, 0x13a0b00, 0x1)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:26 +0x89
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:12 +0xb5
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 28 [runnable]:
sync.(*Map).LoadOrStore(0xc0001cf568, 0x1201100, 0xc00009c000, 0x1220320, 0xc00009a000, 0x1, 0x1, 0x13a0b00)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/map.go:199 +0x6c5
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).loadOrStore(...)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:59
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Lock(0xc0001cf560, 0x13a0b00, 0x1)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:26 +0x89
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:12 +0xb5
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 29 [runnable]:
sync.runtime_SemacquireMutex(0xc0001cf56c, 0xc000039e00, 0x1)
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc0001cf568)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:81
sync.(*Map).LoadOrStore(0xc0001cf568, 0x1201100, 0xc0001cd610, 0x1220320, 0xc0001468c0, 0x1, 0x1, 0x13a0b00)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/map.go:209 +0x60e
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).loadOrStore(...)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:59
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Lock(0xc0001cf560, 0x13a0b00, 0x1)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:26 +0x89
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:12 +0xb5
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 30 [semacquire]:
sync.runtime_SemacquireMutex(0xc0001cf56c, 0xc000220600, 0x1)
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc0001cf568)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:81
sync.(*Map).LoadOrStore(0xc0001cf568, 0x1201100, 0xc0001cd620, 0x1220320, 0xc0001468e0, 0x1, 0x1, 0x13a0b00)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/map.go:209 +0x60e
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).loadOrStore(...)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:59
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Lock(0xc0001cf560, 0x13a0b00, 0x1)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:26 +0x89
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:12 +0xb5
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 31 [semacquire]:
sync.runtime_SemacquireMutex(0xc0001cf56c, 0xc000220e00, 0x1)
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc0001cf568)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:81
sync.(*Map).LoadOrStore(0xc0001cf568, 0x1201100, 0xc0001cd630, 0x1220320, 0xc000146900, 0x1, 0x1, 0x13a0b00)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/map.go:209 +0x60e
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).loadOrStore(...)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:59
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Lock(0xc0001cf560, 0x13a0b00, 0x1)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:26 +0x89
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:12 +0xb5
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc

goroutine 32 [runnable]:
sync.(*Mutex).unlockSlow(0xc0001cf568, 0xc00000002a)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:194 +0xea
sync.(*Mutex).Unlock(...)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/mutex.go:190
sync.(*Map).LoadOrStore(0xc0001cf568, 0x1201100, 0xc000338290, 0x1220320, 0xc000334860, 0x1, 0x3, 0xc000312dc0)
	/usr/local/Cellar/go/1.15.3/libexec/src/sync/map.go:229 +0x2c6
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).loadOrStore(...)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:59
github.com/owncloud/ocis/ocis-pkg/sync.(*NRWMutex).Lock(0xc0001cf560, 0xc000312dc0, 0x3)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex.go:26 +0x89
github.com/owncloud/ocis/ocis-pkg/sync.HammerMutex(0xc0001cf560, 0x7d0, 0xc000102660)
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:12 +0xb5
created by github.com/owncloud/ocis/ocis-pkg/sync.TestNewNRWMutex
	/Users/fschade/Developer/owncloud/ocis/ocis-pkg/sync/nrwmutex_test.go:28 +0xfc
```